[rpc] separate transaction counts rpc into regular vs staking

denniswon opened the pull request:
## Issue

https://github.com/harmony-one/harmony/issues/2759

## Test

launched local network, sent staking and regular txns using hmy cli, from shard 0 with bls key in shard 0. then, saw hmy_getAccountNonce, hmy_getTransactionCount, hmy_getStakingTransactionCount all works correctly. 

```
curl --location --request POST 'http://localhost:9500' --header 'Content-Type: application/json' --data-raw '{
    "jsonrpc": "2.0",
    "method": "hmyv2_getStakingTransactionCount",
    "params": ["one1grlw2qw7t9ulhms38w4ea0gssz0dk8vvsl7jum", "ALL"}],
    "id": 1
}'

curl --location --request POST 'http://localhost:9500' --header 'Content-Type: application/json' --data-raw '{
    "jsonrpc": "2.0",
    "method": "hmyv2_getTransactionCount",
    "params": ["one1grlw2qw7t9ulhms38w4ea0gssz0dk8vvsl7jum", "ALL"}],
    "id": 1
}'

curl --location --request POST 'http://localhost:9500' --header 'Content-Type: application/json' --data-raw '{
    "jsonrpc": "2.0",
    "method": "hmyv2_getAccountNonce",
    "params": ["one1grlw2qw7t9ulhms38w4ea0gssz0dk8vvsl7jum", "latest"}],
    "id": 1
}'
```

```
dennis.won@Jongs-MacBook-Pro:~/harmony-one/harmony (txs_count) $ go test ./node
ok  	github.com/harmony-one/harmony/node	1.674s
```

gupadhyaya commented on hmy/api_backend.go:
```diff
@@ -220,6 +220,16 @@ func (b *APIBackend) GetPoolTransactions() (types.PoolTransactions, error) {
 
+// GetAccountNonce returns the nonce value of the given address for the given block number
+func (b *APIBackend) GetAccountNonce(
+	ctx context.Context, address common.Address, blockNr rpc.BlockNumber) (uint64, error) {
```
No need to worry about the pending transactions anymore?

denniswon replied:
There is getPoolNonce that does the same job

gupadhyaya replied:
I mean to say if there is a pending transaction shouldn't you be returning that number instead of just the committed transactions?

denniswon replied:
ah i see. now understood. yeah, I guess we do need to consider pending txns here.

denniswon replied:
@gupadhyaya actually, isn't pool nonce >= account nonce always? so getPoolNonce() is used for getting number of transactions including pending, and getAccountNonce() used for getting number of confirmed transactions?

gupadhyaya replied:
what will happen if from the cli I made a transaction which is waiting in the pool and then I made the next transaction? Both of them will have the same nonce and only one will work?

gupadhyaya commented on internal/hmyapi/apiv1/transactionpool.go:
```diff
@@ -152,26 +152,6 @@ func (s *PublicTransactionPoolAPI) GetStakingTransactionByBlockHashAndIndex(ctx
 	return nil
 }
 
-// GetTransactionCount returns the number of transactions the given address has sent for the given block number
```
we don't need GetTransactionCount and GetStakingTransactionCount for apiv1?

denniswon replied:
Just deeted the self explanatory comment. Technically dont need to add on v1 but this pr just adds to v1 as well for consistency

denniswon replied:
we actually do. this was why the api_test.sh was failing. Putting this back with comment.

gupadhyaya commented on scripts/api_test.sh:
```diff
@@ -670,7 +670,7 @@ function run_tests() {
 	API_getStorageAt_test
-	API_getTransactionCount_test
+	API_getAccountNonce_test
 	API_sendRawTransaction_test
```
Is it possible to add api tests for transaction and staking transaction counts as well?

denniswon replied:
sure

gupadhyaya commented:
Generally looks good. As, this PR has impact on the clients like js-sdk that uses getTransactionCount for nonce, how do you plan to update the client?

denniswon commented:
@gupadhyaya updated the PR. we actually need to add getTransactionCount api back to v1. Added comment that it is legacy and for the new txns and staking txns counts apis, added them as getTransactionsCount and getStakingTransactionCount.

This was actually api_test.sh was failing. fixed the test by adding the getTransactionCount back to apiv1.

gupadhyaya commented on hmy/api_backend.go:
```diff
@@ -241,6 +251,18 @@ func (b *APIBackend) GetStakingTransactionsHistory(address, txType, order string
+// GetTransactionsCount returns the number of regular transactions of address.
+func (b *APIBackend) GetTransactionsCount(address, txType string) (uint64, error) {
+	count, err := b.hmy.nodeAPI.GetTransactionsCount(address, txType)
+	return count, err
```
simply return `b.hmy.nodeAPI.GetTransactionsCount(address, txType)`

gupadhyaya commented on hmy/api_backend.go:
```diff
@@ -241,6 +251,18 @@ func (b *APIBackend) GetStakingTransactionsHistory(address, txType, order string
+// GetStakingTransactionsCount returns the number of staking transactions of address.
+func (b *APIBackend) GetStakingTransactionsCount(address, txType string) (uint64, error) {
+	count, err := b.hmy.nodeAPI.GetStakingTransactionsCount(address, txType)
+	return count, err
```
same as above.

gupadhyaya commented on internal/hmyapi/apiv1/transactionpool.go:
```diff
@@ -172,6 +174,34 @@ func (s *PublicTransactionPoolAPI) GetTransactionCount(ctx context.Context, addr
+		}
+	}
+	count, err := s.b.GetTransactionsCount(address, txType)
+	return count, err
```
`return s.b.GetTransactionsCount(address, txType)`

gupadhyaya commented on internal/hmyapi/apiv1/transactionpool.go:
```diff
@@ -172,6 +174,34 @@ func (s *PublicTransactionPoolAPI) GetTransactionCount(ctx context.Context, addr
+		}
+	}
+	count, err := s.b.GetStakingTransactionsCount(address, txType)
+	return count, err
```
same as above.

gupadhyaya commented on internal/hmyapi/apiv2/transactionpool.go:
```diff
@@ -200,28 +200,32 @@ func (s *PublicTransactionPoolAPI) GetStakingTransactionByHash(ctx context.Conte
-	if state == nil || err != nil {
-		return 0, err
+	count, err := s.b.GetTransactionsCount(address, txType)
+	return count, err
```
`return s.b.GetTransactionsCount(address, txType)`

gupadhyaya commented on internal/hmyapi/apiv2/transactionpool.go:
```diff
@@ -200,28 +200,32 @@ func (s *PublicTransactionPoolAPI) GetStakingTransactionByHash(ctx context.Conte
-	nonce := state.GetNonce(address)
-	return nonce, state.Error()
+	count, err := s.b.GetStakingTransactionsCount(address, txType)
+	return count, err
```
same as above.

gupadhyaya commented:
@denniswon good to go after fixing the above mentioned minor things.
